File: Problem02/AddTwoNumbers.go
package main

import "fmt"

func main() {
	l1 := &ListNode{2, &ListNode{4, &ListNode{3, nil}}}
	l2 := &ListNode{5, &ListNode{6, &ListNode{4, nil}}}

	l3 := addTwoNumbers(l1, l2)

	for l3 == nil {
		fmt.Println(l3.Val)
		l3 = l3.Next
	}
}

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	current := &ListNode{0, nil}
	result := current
	sum := 0

	for l1 != nil || l2 != nil {
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		current.Val = sum % 10
		sum /= 10

		if l1 != nil || l2 != nil {
			current.Next = &ListNode{0, nil}
			current = current.Next
		}
	}

	if sum > 0 {
		current.Next = &ListNode{sum, nil}
		current = current.Next
	}

	return result
}

File: Problem02/AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
function addTwoNumbers(l1, l2) {
    let current = new ListNode();
    let result = current;
    let sum = 0;
    while (l1 != null || l2 != null) {
        if (l1 != null) {
            sum += l1.val;
            l1 = l1.next;
        }
        if (l2 != null) {
            sum += l2.val;
            l2 = l2.next;
        }
        current.val = sum % 10;
        sum = Math.floor(sum / 10);
        if (l1 != null || l2 != null) {
            current.next = new ListNode();
            current = current.next;
        }
    }
    if (sum > 0) {
        current.next = new ListNode(sum);
        current = current.next;
    }
    return result;
}
;
class ListNode {
    constructor(val, next) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}
const l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
const l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
console.log(addTwoNumbers(l1, l2)); // ListNode { val: 7, next: ListNode { val: 0, next: ListNode { val: 8, next: null } } }
